packages/fx-core/src/component/driver/middleware/addStartAndEndTelemetry.ts

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Middleware, HookContext, NextFunction } from "@feathersjs/hooks/lib";
+import { FxError, Result } from "@microsoft/teamsfx-api";
+import { TeamsFxTelemetryReporter } from "../../utils/teamsFxTelemetryReporter";
+import { DriverContext } from "../interface/commonArgs";
+
+// Based on fx-core's design that a component should always return FxError instead of throw exception, no error handling is added
+export function addStartAndEndTelemetry(eventName: string, componentName: string): Middleware {
+  return async (ctx: HookContext, next: NextFunction) => {
+    const driverContext = ctx.arguments[1] as DriverContext;
+    let telemetryReporter: TeamsFxTelemetryReporter | undefined = undefined;
+    if (driverContext.telemetryReporter) {
+      telemetryReporter = new TeamsFxTelemetryReporter(driverContext.telemetryReporter, {
+        componentName,
+      });
+    }
+    telemetryReporter?.sendStartEvent({ eventName });
+    await next();
+    const result = ctx.result as Result<Map<string, string>, FxError>;
+    if (result.isOk()) {
+      telemetryReporter?.sendEndEvent({ eventName });
```

Comment:
When we send end telemetry, we may set customized properties for it. It is hard to set customized properties in this method interface.

Reply:
since we don't have such requirement yet, I didn't add this feature. If we need such feature in the future, the change is simiple:
1. Add a new optional properties parameter: `addStartAndEndTelemetry(eventName: string, componentName: string, properties?: { [key:string}:string })`
2. Pass properties to `sendEnvEvent`: `telemetryReporter?.sendEnvEvent({ eventName, properties });`. The underlying reporter will automatically merge the properties.

How do you think of keeping things simple for now?

Reply:
Yeah. It is fine to keep things simple. It is good now.
By the way, for the future update, as the customized properties are always generated in the runtime, we may figure out how to get these properties.